proof/refine/RISCV64/KHeap_R.thy

```diff
@@ -2105,32 +3034,1807 @@ lemma doMachineOp_invs_bits[wp]:
+
+lemma update_sched_context_rewrite:
+  "monadic_rewrite False True (sc_obj_at n scp)
+    (update_sched_context scp f)
+      (do sc \<leftarrow> get_sched_context scp;
+          set_object scp (kernel_object.SchedContext (f sc) n) od)"
+  apply (clarsimp simp: update_sched_context_def get_sched_context_def bind_assoc)
+  apply (rule monadic_rewrite_bind_tail)
+   defer
+   apply (rule get_object_sp)
+  apply (case_tac obj;
+         fastforce simp: monadic_rewrite_refl3 set_object_def monadic_rewrite_def obj_at_def
+                         is_sc_obj_def)
+  done
+
+lemma updateSchedContext_corres_gen:
+  assumes
+      R1: "\<forall>s s'. (s, s') \<in> state_relation \<longrightarrow>
+           P s \<longrightarrow> P' s' \<longrightarrow> sc_at ptr s \<longrightarrow> sc_at' ptr s' \<longrightarrow>
+           (\<forall>n. (((\<lambda>ko. obj_bits ko = min_sched_context_bits + n) |< kheap s) ptr)\<longrightarrow>
+           sc_relation (the ((scs_of2 s ||> f) ptr)) n (the ((scs_of' s' ||> f') ptr)))"
+  and R2: "\<forall>s s'. (s, s') \<in> state_relation \<longrightarrow>
+          P s \<longrightarrow> P' s' \<longrightarrow> sc_at ptr s \<longrightarrow> sc_at' ptr s' \<longrightarrow>
+           heap_ls (replyPrevs_of s')  (scReply (the ((scs_of' s' ||> f') ptr)))
+             (sc_replies (the ((scs_of2 s ||> f) ptr)))"
+  and sz: "\<forall>sc'::sched_context. objBits sc' = objBits (f' sc')"
+  shows "corres dc
+         (sc_at ptr and P)
+         (sc_at' ptr and P')
+            (update_sched_context ptr f)
+            (updateSchedContext ptr f')"
+  unfolding corres_underlying_def using sz
+  apply clarsimp
+  apply (rename_tac s s')
+  apply (drule obj_at_ko_at)
+  apply (drule obj_at_ko_at')
+  apply (clarsimp simp: is_sc_obj)
+  apply (rename_tac sc' n sc)
+  apply (rule conjI, clarsimp)
+   apply (erule use_valid[OF _ updateSchedContext_wp])
+   apply clarsimp
+   apply (rule_tac x="((), s\<lparr>kheap := kheap s(ptr \<mapsto>
+                  kernel_object.SchedContext (f sc) n)\<rparr>)" in bexI)
+    apply clarsimp
+    apply (drule state_relation_sc_update[OF R1 R2 sz, simplified])
+      apply ((fastforce simp: obj_at_def is_sc_obj obj_at'_def projectKOs)+)[4]
+    apply (clarsimp simp: obj_at_def obj_at'_def projectKOs fun_upd_def[symmetric] opt_map_red)
+    apply (case_tac s; case_tac s'; fastforce)
+   apply (clarsimp simp: update_sched_context_def obj_at_def in_monad
+                         get_object_def set_object_def a_type_def)
+  apply (clarsimp intro!: no_failD[OF no_fail_updateSchedContext]
+                    simp: obj_at'_def projectKOs opt_map_def)
+  done
+
+lemmas updateSchedContext_corres = updateSchedContext_corres_gen[where P=\<top> and P'=\<top>, simplified]
+
+(* end : updateSchedContext *)
+
+end
+
+(* FIXME RT: rename *)
+(* this lets cross the sc size information from concrete to abstract *)
+lemma ko_at'_cross:
```

Comment:
We should perform the rename to something like `ko_at_sc_cross` before we perpetuate the name. Might be relatively simple to do on ARM as well.